File: src/python/xerxes/channels/registry.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Channel registry and lifecycle management.

``ChannelRegistry`` is the daemon-facing collection that owns a process's
channels: it gives each channel a name, holds the inbound handler, and
runs ``start_all``/``stop_all`` with best-effort error handling so a
single misconfigured adapter cannot prevent the others from coming up.
``gather_inbound`` is a thin coroutine helper for starting several
registries concurrently.
"""

from __future__ import annotations

import asyncio
import logging
import typing as tp

from .base import Channel
from .types import ChannelMessage

logger = logging.getLogger(__name__)
InboundHandler = tp.Callable[[ChannelMessage], tp.Awaitable[None]]


class ChannelRegistry:
    """Named collection of channels with start/stop lifecycle bookkeeping.

    Tracks which channels have been started so ``start_all`` is idempotent
    and ``stop_all`` only stops what is actually running. Failures during
    start are logged but do not abort the rest of the batch — operators can
    misconfigure one adapter without losing all messaging.
    """

    def __init__(self) -> None:
        """Build an empty registry with no inbound handler set."""
        self._channels: dict[str, Channel] = {}
        self._handler: InboundHandler | None = None
        self._started: set[str] = set()

    def register(self, name: str, channel: Channel) -> None:
        """Add or replace a channel under ``name``.

        Args:
            name: Registry-level identifier. Conventionally matches
                ``channel.name``, but the registry does not enforce it.
            channel: Channel instance to register.
        """
        self._channels[name] = channel

    def unregister(self, name: str) -> None:
        """Remove a channel and forget its started state.

        Does not stop the channel — callers must do that themselves before
        unregistering if they want a clean shutdown.

        Args:
            name: Identifier passed to ``register``.
        """
        self._channels.pop(name, None)
        self._started.discard(name)

    def get(self, name: str) -> Channel | None:
        """Look up a channel by name.

        Args:
            name: Identifier passed to ``register``.

        Returns:
            The channel if registered, otherwise ``None``.
        """
        return self._channels.get(name)

    def all(self) -> dict[str, Channel]:
        """Return a shallow copy of the ``name → channel`` mapping.

        Mutating the returned dict does not affect the registry.
        """
        return dict(self._channels)

    def names(self) -> list[str]:
        """Return the registry's channel names in insertion order."""
        return list(self._channels.keys())

    def set_handler(self, handler: InboundHandler) -> None:
        """Install the single inbound callback used by every channel.

        Must be called before ``start_all`` — channels share one handler so
        the registry can route all inbound traffic into a single agent
        runtime regardless of source platform.

        Args:
            handler: Async callable invoked for every inbound message.
        """
        self._handler = handler

    async def start_all(self) -> None:
        """Start every registered channel that is not already running.

        Each ``Channel.start`` is awaited inside a try/except so one bad
        adapter does not prevent the others from coming up; failures are
        logged at WARNING with the traceback.

        Raises:
            RuntimeError: ``set_handler`` was not called first.
        """
        if self._handler is None:
            raise RuntimeError("ChannelRegistry.set_handler must be called before start_all()")
        for name, channel in list(self._channels.items()):
            if name in self._started:
                continue
            try:
                await channel.start(self._handler)
                self._started.add(name)
                logger.info("Channel started: %s", name)
            except Exception:
                logger.warning("Failed to start channel %s", name, exc_info=True)

    async def stop_all(self) -> None:
        """Stop every currently running channel; swallow per-channel errors.

        Channels whose ``stop`` raises are logged at DEBUG and still removed
        from the started set so the registry remains consistent.
        """
        for name in list(self._started):
            channel = self._channels.get(name)
            if channel is None:
                self._started.discard(name)
                continue
            try:
                await channel.stop()
            except Exception:
                logger.debug("Channel %s raised on stop", name, exc_info=True)
            self._started.discard(name)

    async def send(self, message: ChannelMessage) -> None:
        """Route an outbound message to the channel named by ``message.channel``.

        Args:
            message: Outbound message; ``message.channel`` selects the target.

        Raises:
            KeyError: No channel is registered under ``message.channel``.
        """
        chan = self._channels.get(message.channel)
        if chan is None:
            raise KeyError(f"unknown channel {message.channel!r}")
        await chan.send(message)

    def discover_entry_points(self, group: str = "xerxes.channels") -> list[str]:
        """Auto-register channels declared via ``importlib.metadata`` entry points.

        Each entry point must be a zero-arg factory returning a ``Channel``
        instance. Factories that fail to import or return a non-``Channel``
        are skipped with a logged warning so a single broken plugin cannot
        crash startup.

        Args:
            group: Entry-point group name. Defaults to ``"xerxes.channels"``.

        Returns:
            Names of channels that were loaded and registered successfully.
        """
        try:
            from importlib.metadata import entry_points
        except Exception:
            return []
        added: list[str] = []
        try:
            eps = entry_points(group=group)
        except TypeError:
            eps = entry_points().get(group, tp.cast(tp.Any, []))
        for ep in eps:
            try:
                factory = ep.load()
                chan = factory()
                if isinstance(chan, Channel):
                    self.register(ep.name, chan)
                    added.append(ep.name)
            except Exception:
                logger.warning("Channel entry-point %s failed to load", ep.name, exc_info=True)
        return added


def gather_inbound(*registries: ChannelRegistry) -> tp.Awaitable[None]:
    """Start ``start_all`` on every registry concurrently.

    Args:
        *registries: One or more registries whose channels should come up
            in parallel.

    Returns:
        A coroutine that resolves when every registry finishes its
        ``start_all`` (including channels that failed to start individually).
    """

    async def _run():
        """Run ``start_all`` on every registry concurrently via ``asyncio.gather``."""
        await asyncio.gather(*(r.start_all() for r in registries))

    return _run()


__all__ = ["ChannelRegistry", "InboundHandler", "gather_inbound"]

File: xerxes/src/channels/registry.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import type { Channel, InboundHandler } from './base.js'
import type { ChannelMessage } from './types.js'

export type ChannelLifecycleOperation = 'start' | 'stop'

/** A contained channel lifecycle failure available to daemon diagnostics. */
export interface ChannelLifecycleFailure {
  readonly channel: string
  readonly error: unknown
  readonly operation: ChannelLifecycleOperation
}

export interface ChannelRegistryOptions {
  /** Receives failures that are intentionally isolated from other channels. */
  readonly onFailure?: (failure: ChannelLifecycleFailure) => void
}

/** Raised when an outbound message names a channel that is not registered. */
export class UnknownChannelError extends Error {
  readonly channel: string

  constructor(channel: string) {
    super(`unknown channel '${channel}'`)
    this.name = new.target.name
    this.channel = channel
  }
}

/**
 * Named collection of channels with lifecycle bookkeeping.
 *
 * A failed adapter never prevents healthy adapters from starting or stopping.
 * Failures are retained for status reporting and can be forwarded to the
 * daemon's logger through `onFailure`.
 */
export class ChannelRegistry {
  private readonly channels = new Map<string, Channel>()
  private readonly failures: ChannelLifecycleFailure[] = []
  private handler: InboundHandler | undefined
  private lifecycle: Promise<void> = Promise.resolve()
  private readonly onFailure:
    ((failure: ChannelLifecycleFailure) => void) | undefined
  private readonly started = new Map<string, Channel>()

  constructor(options: ChannelRegistryOptions = {}) {
    this.onFailure = options.onFailure
  }

  /** Add or replace a channel under its daemon-facing name. */
  register(name: string, channel: Channel): void {
    this.channels.set(name, channel)
  }

  /** Remove a channel without stopping it; callers control teardown explicitly. */
  unregister(name: string): void {
    this.channels.delete(name)
    this.started.delete(name)
  }

  get(name: string): Channel | undefined {
    return this.channels.get(name)
  }

  /** Return an isolated copy of the current name-to-channel mapping. */
  all(): Map<string, Channel> {
    return new Map(this.channels)
  }

  names(): string[] {
    return [...this.channels.keys()]
  }

  startedNames(): string[] {
    return [...this.started.keys()]
  }

  isStarted(name: string): boolean {
    return this.started.has(name)
  }

  /** Install the single inbound callback shared by every registered channel. */
  setHandler(handler: InboundHandler): void {
    this.handler = handler
  }

  hasHandler(): boolean {
    return this.handler !== undefined
  }

  /** Return a snapshot of contained lifecycle failures. */
  lifecycleFailures(): readonly ChannelLifecycleFailure[] {
    return [...this.failures]
  }

  clearLifecycleFailures(name?: string): void {
    if (name === undefined) {
      this.failures.length = 0
      return
    }
    let target = 0
    for (const failure of this.failures) {
      if (failure.channel !== name) {
        this.failures[target] = failure
        target += 1
      }
    }
    this.failures.length = target
  }

  /** Start one registered channel through the shared inbound handler. */
  async start(name: string): Promise<void> {
    return this.enqueue(async () => {
      const channel = this.requireChannel(name)
      const handler = this.requireHandler('start')
      await this.startOne(name, channel, handler, true)
    })
  }

  /** Stop one registered channel without affecting other enabled adapters. */
  async stop(name: string): Promise<void> {
    return this.enqueue(async () => {
      const channel = this.requireChannel(name)
      await this.stopOne(name, channel, true)
    })
  }

  /** Start every registered channel that is not already running. */
  async startAll(): Promise<void> {
    return this.enqueue(async () => {
      const handler = this.requireHandler('startAll')
      for (const [name, channel] of [...this.channels]) {
        await this.startOne(name, channel, handler, false)
      }
    })
  }

  /** Stop every running channel while isolating teardown failures. */
  async stopAll(): Promise<void> {
    return this.enqueue(async () => {
      for (const [name, channel] of [...this.started]) {
        await this.stopOne(name, channel, false)
      }
    })
  }

  /** Route an outbound message using its normalized channel name. */
  async send(message: ChannelMessage): Promise<void> {
    const channel = this.channels.get(message.channel)
    if (!channel) {
      throw new UnknownChannelError(message.channel)
    }
    await channel.send(message)
  }

  private enqueue(operation: () => Promise<void>): Promise<void> {
    const pending = this.lifecycle.then(operation, operation)
    this.lifecycle = pending.then(
      () => undefined,
      () => undefined,
    )
    return pending
  }

  private requireChannel(name: string): Channel {
    const channel = this.channels.get(name)
    if (!channel) {
      throw new UnknownChannelError(name)
    }
    return channel
  }

  private requireHandler(operation: 'start' | 'startAll'): InboundHandler {
    if (!this.handler) {
      throw new Error(operation === 'startAll'
        ? 'ChannelRegistry.setHandler must be called before startAll()'
        : 'ChannelRegistry.setHandler must be called before starting a channel')
    }
    return this.handler
  }

  private async startOne(
    name: string,
    channel: Channel,
    handler: InboundHandler,
    propagateFailure: boolean,
  ): Promise<void> {
    const running = this.started.get(name)
    if (running === channel) {
      return
    }
    if (running) {
      try {
        await running.stop()
        this.started.delete(name)
      } catch (error) {
        this.report({ channel: name, error, operation: 'stop' })
        if (propagateFailure) {
          throw error
        }
        return
      }
    }
    try {
      await channel.start(handler)
      this.started.set(name, channel)
    } catch (error) {
      this.report({ channel: name, error, operation: 'start' })
      if (propagateFailure) {
        throw error
      }
    }
  }

  private async stopOne(name: string, channel: Channel, propagateFailure: boolean): Promise<void> {
    try {
      await channel.stop()
    } catch (error) {
      this.report({ channel: name, error, operation: 'stop' })
      if (propagateFailure) {
        throw error
      }
    } finally {
      this.started.delete(name)
    }
  }

  private report(failure: ChannelLifecycleFailure): void {
    this.failures.push(failure)
    if (!this.onFailure) {
      return
    }
    try {
      this.onFailure(failure)
    } catch {
      // Reporting must not make an isolated adapter failure abort the batch.
    }
  }
}

/** Start multiple registries concurrently, matching the daemon's inbound fan-in. */
export async function gatherInbound(
  ...registries: readonly ChannelRegistry[]
): Promise<void> {
  await Promise.all(registries.map((registry) => registry.startAll()))
}
